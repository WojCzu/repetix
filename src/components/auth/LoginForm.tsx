import { useState, type ChangeEvent, type FormEvent } from "react";
import { Button } from "../ui/button";
import { Input } from "../ui/input";
import { Label } from "../ui/label";
import { ValidationMessage } from "../ui/ValidationMessage";
import { Card, CardContent, CardFooter, CardHeader, CardTitle } from "../ui/card";
import { Link } from "../ui/link";
import { loginSchema, type LoginFormData } from "@/lib/schemas/auth.schema";

interface LoginFormProps {
  onSubmit: (data: LoginFormData) => Promise<void>;
  redirectTo?: string;
}

export function LoginForm({ onSubmit, redirectTo }: LoginFormProps) {
  const [formData, setFormData] = useState<LoginFormData>({
    email: "",
    password: "",
  });

  const [errors, setErrors] = useState<Record<string, string>>({});
  const [isLoading, setIsLoading] = useState(false);

  const handleChange = (field: keyof LoginFormData) => (e: ChangeEvent<HTMLInputElement>) => {
    setFormData((prev) => ({ ...prev, [field]: e.target.value }));
  };

  const validateForm = () => {
    const result = loginSchema.safeParse(formData);

    if (!result.success) {
      const validationErrors: Record<string, string> = {};
      result.error.errors.forEach((err) => {
        const field = err.path[0];
        if (field) {
          validationErrors[field] = err.message;
        }
      });
      setErrors(validationErrors);
      return false;
    }

    setErrors({});
    return true;
  };

  const handleSubmit = async (e: FormEvent) => {
    e.preventDefault();

    if (!validateForm()) return;

    setIsLoading(true);
    try {
      await onSubmit(formData);
      if (redirectTo) {
        window.location.href = redirectTo;
      }
    } catch (error) {
      if (error instanceof Error) {
        setErrors({ submit: error.message });
      }
    } finally {
      setIsLoading(false);
    }
  };

  return (
    <Card>
      <CardHeader>
        <CardTitle>Sign in to your account</CardTitle>
      </CardHeader>
      <form onSubmit={handleSubmit}>
        <CardContent className="space-y-6">
          <div className="space-y-2">
            <Label htmlFor="email">Email</Label>
            <Input
              id="email"
              type="email"
              value={formData.email}
              onChange={handleChange("email")}
              placeholder="name@example.com"
              required
              aria-describedby={errors.email ? "email-error" : undefined}
            />
            {errors.email && <ValidationMessage id="email-error" message={errors.email} />}
          </div>

          <div className="space-y-2">
            <Label htmlFor="password">Password</Label>
            <Input
              id="password"
              type="password"
              value={formData.password}
              onChange={handleChange("password")}
              required
              aria-describedby={errors.password ? "password-error" : undefined}
            />
            {errors.password && <ValidationMessage id="password-error" message={errors.password} />}
          </div>

          {errors.submit && <ValidationMessage id="submit-error" message={errors.submit} />}
        </CardContent>

        <CardFooter className="flex flex-col space-y-4 pt-6">
          <Button type="submit" className="w-full" disabled={isLoading}>
            {isLoading ? "Signing in..." : "Sign in"}
          </Button>
          <div className="flex flex-col space-y-2 text-center text-sm text-muted-foreground">
            <p>
              Don&apos;t have an account?{" "}
              <Link href="/register" className="text-primary hover:underline">
                Create an account
              </Link>
            </p>
            <p>
              <Link href="/reset-password" className="text-primary hover:underline">
                Forgot your password?
              </Link>
            </p>
          </div>
        </CardFooter>
      </form>
    </Card>
  );
}
