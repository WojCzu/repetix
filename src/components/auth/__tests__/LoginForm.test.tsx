import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { LoginForm } from "../LoginForm";
import { axe } from "../../../test/vitest.setup";
import type { LoginFormData } from "@/lib/schemas/auth.schema";

// Mock zewnętrznych zależności
vi.mock("@/lib/schemas/auth.schema", async (importOriginal) => {
  const actual = await importOriginal<typeof import("@/lib/schemas/auth.schema")>();
  return {
    ...actual,
    loginSchema: {
      parse: vi.fn((data: LoginFormData) => {
        const errors: { errors?: { path: string[]; message: string }[] } = { errors: [] };

        if (!data.email) {
          errors.errors?.push({ path: ["email"], message: "Email is required" });
        } else if (!data.email.includes("@")) {
          errors.errors?.push({ path: ["email"], message: "Invalid email format" });
        }

        if (!data.password) {
          errors.errors?.push({ path: ["password"], message: "Password is required" });
        } else if (data.password.length < 8) {
          errors.errors?.push({
            path: ["password"],
            message: "Password must be at least 8 characters long",
          });
        }

        if (errors.errors?.length) {
          const error = new Error("Validation failed");
          Object.assign(error, errors);
          throw error;
        }

        return data;
      }),
    },
  };
});

describe("LoginForm", () => {
  const mockOnSubmit = vi.fn();
  let user: ReturnType<typeof userEvent.setup>;

  beforeEach(() => {
    user = userEvent.setup();
    mockOnSubmit.mockReset();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  describe("Rendering", () => {
    it("renders all required form elements", () => {
      render(<LoginForm onSubmit={mockOnSubmit} />);

      expect(screen.getByText(/sign in to your account/i)).toBeInTheDocument();
      expect(screen.getByLabelText(/email/i)).toBeInTheDocument();
      expect(screen.getByLabelText(/password/i)).toBeInTheDocument();
      expect(screen.getByRole("button", { name: /sign in/i })).toBeInTheDocument();
      expect(screen.getByText(/don't have an account/i)).toBeInTheDocument();
      expect(screen.getByText(/forgot your password/i)).toBeInTheDocument();
    });

    it("renders with proper ARIA attributes for accessibility", () => {
      render(<LoginForm onSubmit={mockOnSubmit} />);

      const emailInput = screen.getByLabelText(/email/i);
      const passwordInput = screen.getByLabelText(/password/i);

      expect(emailInput).toHaveAttribute("aria-invalid", "false");
      expect(emailInput).not.toHaveAttribute("aria-describedby");
      expect(passwordInput).toHaveAttribute("aria-invalid", "false");
      expect(passwordInput).not.toHaveAttribute("aria-describedby");
    });

    it("disables form elements when isLoading is true", () => {
      render(<LoginForm onSubmit={mockOnSubmit} isLoading={true} />);

      expect(screen.getByLabelText(/email/i)).toBeDisabled();
      expect(screen.getByLabelText(/password/i)).toBeDisabled();
      expect(screen.getByRole("button", { name: /signing in/i })).toBeDisabled();
    });

    it("changes button text when isLoading is true", () => {
      render(<LoginForm onSubmit={mockOnSubmit} isLoading={true} />);

      expect(screen.getByRole("button", { name: /signing in/i })).toBeInTheDocument();
      expect(screen.queryByRole("button", { name: /sign in$/i })).not.toBeInTheDocument();
    });

    it("has no accessibility violations", async () => {
      const { container } = render(<LoginForm onSubmit={mockOnSubmit} />);

      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  });

  describe("Form Validation", () => {
    it("shows validation error for empty email", async () => {
      render(<LoginForm onSubmit={mockOnSubmit} />);

      const submitButton = screen.getByRole("button", { name: /sign in$/i });
      await user.click(submitButton);

      expect(await screen.findByText(/email is required/i)).toBeInTheDocument();
      expect(mockOnSubmit).not.toHaveBeenCalled();
    });

    it("shows validation error for invalid email format", async () => {
      render(<LoginForm onSubmit={mockOnSubmit} />);

      const emailInput = screen.getByLabelText(/email/i);
      await user.type(emailInput, "invalid-email");

      const submitButton = screen.getByRole("button", { name: /sign in$/i });
      await user.click(submitButton);

      expect(await screen.findByText(/invalid email format/i)).toBeInTheDocument();
      expect(mockOnSubmit).not.toHaveBeenCalled();
    });

    it("shows validation error for empty password", async () => {
      render(<LoginForm onSubmit={mockOnSubmit} />);

      const emailInput = screen.getByLabelText(/email/i);
      await user.type(emailInput, "test@example.com");

      const submitButton = screen.getByRole("button", { name: /sign in$/i });
      await user.click(submitButton);

      expect(await screen.findByText(/password is required/i)).toBeInTheDocument();
      expect(mockOnSubmit).not.toHaveBeenCalled();
    });

    it("shows validation error for too short password", async () => {
      render(<LoginForm onSubmit={mockOnSubmit} />);

      const emailInput = screen.getByLabelText(/email/i);
      await user.type(emailInput, "test@example.com");

      const passwordInput = screen.getByLabelText(/password/i);
      await user.type(passwordInput, "short");

      const submitButton = screen.getByRole("button", { name: /sign in$/i });
      await user.click(submitButton);

      expect(await screen.findByText(/password must be at least 8 characters long/i)).toBeInTheDocument();
      expect(mockOnSubmit).not.toHaveBeenCalled();
    });

    it("clears field-specific error when user starts typing", async () => {
      render(<LoginForm onSubmit={mockOnSubmit} />);

      // Trigger validation error
      const submitButton = screen.getByRole("button", { name: /sign in$/i });
      await user.click(submitButton);

      // Verify error is shown
      const emailError = await screen.findByText(/email is required/i);
      expect(emailError).toBeInTheDocument();

      // Type in the field
      const emailInput = screen.getByLabelText(/email/i);
      await user.type(emailInput, "a");

      // Verify error is cleared
      await waitFor(() => {
        expect(screen.queryByText(/email is required/i)).not.toBeInTheDocument();
      });
    });
  });

  describe("Form Submission", () => {
    it("calls onSubmit with correct data on valid form submission", async () => {
      render(<LoginForm onSubmit={mockOnSubmit} />);

      const emailInput = screen.getByLabelText(/email/i);
      const passwordInput = screen.getByLabelText(/password/i);
      const submitButton = screen.getByRole("button", { name: /sign in$/i });

      await user.type(emailInput, "test@example.com");
      await user.type(passwordInput, "password123");
      await user.click(submitButton);

      await waitFor(() => {
        expect(mockOnSubmit).toHaveBeenCalledTimes(1);
        expect(mockOnSubmit).toHaveBeenCalledWith({
          email: "test@example.com",
          password: "password123",
        });
      });
    });

    it("displays API error message when onSubmit throws an error", async () => {
      mockOnSubmit.mockRejectedValueOnce(new Error("Invalid login credentials"));

      render(<LoginForm onSubmit={mockOnSubmit} />);

      const emailInput = screen.getByLabelText(/email/i);
      const passwordInput = screen.getByLabelText(/password/i);
      const submitButton = screen.getByRole("button", { name: /sign in$/i });

      await user.type(emailInput, "test@example.com");
      await user.type(passwordInput, "password123");
      await user.click(submitButton);

      // Sprawdzamy nie tylko czy element alert istnieje, ale również jego zawartość
      const errorMessage = await screen.findByRole("alert");
      expect(errorMessage).toHaveTextContent(/invalid email or password/i);
    });

    it("handles network error", async () => {
      mockOnSubmit.mockRejectedValueOnce(new Error("network"));

      render(<LoginForm onSubmit={mockOnSubmit} />);

      const emailInput = screen.getByLabelText(/email/i);
      const passwordInput = screen.getByLabelText(/password/i);
      const submitButton = screen.getByRole("button", { name: /sign in$/i });

      await user.type(emailInput, "test@example.com");
      await user.type(passwordInput, "password123");
      await user.click(submitButton);

      // Sprawdzamy dokładną treść komunikatu błędu
      const errorMessage = await screen.findByRole("alert");
      expect(errorMessage).toHaveTextContent(/network error/i);
    });

    it("disables form during submission", async () => {
      // Create a delayed promise to check the form state during submission
      let resolveSubmit: ((value: unknown) => void) | undefined;
      mockOnSubmit.mockImplementationOnce(
        () =>
          new Promise((resolve) => {
            resolveSubmit = resolve;
          })
      );

      render(<LoginForm onSubmit={mockOnSubmit} />);

      const emailInput = screen.getByLabelText(/email/i);
      const passwordInput = screen.getByLabelText(/password/i);
      const submitButton = screen.getByRole("button", { name: /sign in$/i });

      await user.type(emailInput, "test@example.com");
      await user.type(passwordInput, "password123");

      // Submit form
      const submitPromise = user.click(submitButton);

      // Check form state during submission
      await waitFor(() => {
        expect(emailInput).toBeDisabled();
        expect(passwordInput).toBeDisabled();
        expect(submitButton).toBeDisabled();
      });

      // Resolve the submission
      if (resolveSubmit) {
        resolveSubmit({});
      }
      await submitPromise;

      // Ensure form is enabled again
      await waitFor(() => {
        expect(emailInput).not.toBeDisabled();
        expect(passwordInput).not.toBeDisabled();
        expect(submitButton).not.toBeDisabled();
      });
    });

    it("passes correct submission state to UI", () => {
      // Test dla isLoading=true
      const { rerender } = render(<LoginForm onSubmit={mockOnSubmit} isLoading={true} />);

      const loadingButton = screen.getByRole("button", { name: /signing in/i });
      expect(loadingButton).toBeDisabled();

      // Test dla isLoading=false
      rerender(<LoginForm onSubmit={mockOnSubmit} isLoading={false} />);

      const enabledButton = screen.getByRole("button", { name: /sign in$/i });
      expect(enabledButton).not.toBeDisabled();
    });
  });

  describe("Edge Cases", () => {
    it("handles non-Error exceptions during form submission", async () => {
      mockOnSubmit.mockRejectedValueOnce("Unknown error");

      render(<LoginForm onSubmit={mockOnSubmit} />);

      const emailInput = screen.getByLabelText(/email/i);
      const passwordInput = screen.getByLabelText(/password/i);
      const submitButton = screen.getByRole("button", { name: /sign in$/i });

      await user.type(emailInput, "test@example.com");
      await user.type(passwordInput, "password123");
      await user.click(submitButton);

      // Sprawdzamy dokładną treść komunikatu błędu
      const errorMessage = await screen.findByRole("alert");
      expect(errorMessage).toHaveTextContent(/an unexpected error occurred/i);
    });

    it("prevents submitting multiple times during an ongoing submission", async () => {
      // Create a promise that won't resolve during the test
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      mockOnSubmit.mockImplementationOnce(() => new Promise(() => {}));

      render(<LoginForm onSubmit={mockOnSubmit} />);

      const emailInput = screen.getByLabelText(/email/i);
      const passwordInput = screen.getByLabelText(/password/i);
      const submitButton = screen.getByRole("button", { name: /sign in$/i });

      await user.type(emailInput, "test@example.com");
      await user.type(passwordInput, "password123");

      // First click
      await user.click(submitButton);

      // Second click attempt (should be prevented by disabled state)
      await user.click(submitButton);

      expect(mockOnSubmit).toHaveBeenCalledTimes(1);
    });

    it("sanitizes inputs by trimming whitespace", async () => {
      render(<LoginForm onSubmit={mockOnSubmit} />);

      const emailInput = screen.getByLabelText(/email/i);
      const passwordInput = screen.getByLabelText(/password/i);
      const submitButton = screen.getByRole("button", { name: /sign in$/i });

      // Wprowadzamy dane z białymi znakami na początku i końcu
      await user.type(emailInput, " test@example.com ");
      await user.type(passwordInput, " password123 ");
      await user.click(submitButton);

      // Weryfikujemy, że do funkcji onSubmit zostały przekazane dane bez białych znaków
      await waitFor(() => {
        expect(mockOnSubmit).toHaveBeenCalledWith({
          email: "test@example.com",
          password: "password123",
        });
      });
    });
  });

  describe("Snapshot Tests", () => {
    it("matches snapshot in default state", () => {
      const { asFragment } = render(<LoginForm onSubmit={mockOnSubmit} />);
      expect(asFragment()).toMatchSnapshot();
    });

    it("matches snapshot in loading state", () => {
      const { asFragment } = render(<LoginForm onSubmit={mockOnSubmit} isLoading={true} />);
      expect(asFragment()).toMatchSnapshot();
    });

    it("matches snapshot with validation errors", async () => {
      const { asFragment } = render(<LoginForm onSubmit={mockOnSubmit} />);

      const submitButton = screen.getByRole("button", { name: /sign in$/i });
      await user.click(submitButton);

      // Wait for validation errors to appear
      await screen.findByText(/email is required/i);

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
