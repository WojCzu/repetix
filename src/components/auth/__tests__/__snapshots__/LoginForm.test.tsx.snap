// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginForm > Snapshot Tests > matches snapshot in default state 1`] = `
<DocumentFragment>
  <div
    class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
    data-slot="card"
  >
    <div
      class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6"
      data-slot="card-header"
    >
      <div
        class="leading-none font-semibold"
        data-slot="card-title"
      >
        Sign in to your account
      </div>
    </div>
    <form
      data-testid="login-form"
      novalidate=""
    >
      <div
        class="px-6 space-y-6"
        data-slot="card-content"
      >
        <div
          class="space-y-2"
        >
          <label
            class="flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50"
            data-slot="label"
            for="email"
          >
            Email
          </label>
          <input
            aria-invalid="false"
            class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive"
            data-slot="input"
            data-testid="email-input"
            id="email"
            placeholder="name@example.com"
            required=""
            type="email"
            value=""
          />
        </div>
        <div
          class="space-y-2"
        >
          <label
            class="flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50"
            data-slot="label"
            for="password"
          >
            Password
          </label>
          <input
            aria-invalid="false"
            class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive"
            data-slot="input"
            data-testid="password-input"
            id="password"
            required=""
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        class="items-center px-6 [.border-t]:pt-6 flex flex-col space-y-4 pt-6"
        data-slot="card-footer"
      >
        <button
          class="inline-flex items-center justify-center cursor-pointer gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3 w-full"
          data-slot="button"
          data-testid="submit-button"
          type="submit"
        >
          Sign in
        </button>
        <div
          class="flex flex-col space-y-2 text-center text-sm text-muted-foreground"
        >
          <p>
            Don't have an account? 
            <a
              class="inline-flex items-center justify-center cursor-pointer gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive underline-offset-4 h-9 px-4 py-2 has-[>svg]:px-3 text-primary hover:underline"
              data-testid="register-link"
              href="/register"
            >
              Create an account
            </a>
          </p>
          <p>
            <a
              class="inline-flex items-center justify-center cursor-pointer gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive underline-offset-4 h-9 px-4 py-2 has-[>svg]:px-3 text-primary hover:underline"
              data-testid="reset-password-link"
              href="/reset-password"
            >
              Forgot your password?
            </a>
          </p>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`LoginForm > Snapshot Tests > matches snapshot in loading state 1`] = `
<DocumentFragment>
  <div
    class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
    data-slot="card"
  >
    <div
      class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6"
      data-slot="card-header"
    >
      <div
        class="leading-none font-semibold"
        data-slot="card-title"
      >
        Sign in to your account
      </div>
    </div>
    <form
      data-testid="login-form"
      novalidate=""
    >
      <div
        class="px-6 space-y-6"
        data-slot="card-content"
      >
        <div
          class="space-y-2"
        >
          <label
            class="flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50"
            data-slot="label"
            for="email"
          >
            Email
          </label>
          <input
            aria-invalid="false"
            class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive"
            data-slot="input"
            data-testid="email-input"
            disabled=""
            id="email"
            placeholder="name@example.com"
            required=""
            type="email"
            value=""
          />
        </div>
        <div
          class="space-y-2"
        >
          <label
            class="flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50"
            data-slot="label"
            for="password"
          >
            Password
          </label>
          <input
            aria-invalid="false"
            class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive"
            data-slot="input"
            data-testid="password-input"
            disabled=""
            id="password"
            required=""
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        class="items-center px-6 [.border-t]:pt-6 flex flex-col space-y-4 pt-6"
        data-slot="card-footer"
      >
        <button
          class="inline-flex items-center justify-center cursor-pointer gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3 w-full"
          data-slot="button"
          data-testid="submit-button"
          disabled=""
          type="submit"
        >
          Signing in...
        </button>
        <div
          class="flex flex-col space-y-2 text-center text-sm text-muted-foreground"
        >
          <p>
            Don't have an account? 
            <a
              class="inline-flex items-center justify-center cursor-pointer gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive underline-offset-4 h-9 px-4 py-2 has-[>svg]:px-3 text-primary hover:underline"
              data-testid="register-link"
              href="/register"
            >
              Create an account
            </a>
          </p>
          <p>
            <a
              class="inline-flex items-center justify-center cursor-pointer gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive underline-offset-4 h-9 px-4 py-2 has-[>svg]:px-3 text-primary hover:underline"
              data-testid="reset-password-link"
              href="/reset-password"
            >
              Forgot your password?
            </a>
          </p>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`LoginForm > Snapshot Tests > matches snapshot with validation errors 1`] = `
<DocumentFragment>
  <div
    class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
    data-slot="card"
  >
    <div
      class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6"
      data-slot="card-header"
    >
      <div
        class="leading-none font-semibold"
        data-slot="card-title"
      >
        Sign in to your account
      </div>
    </div>
    <form
      data-testid="login-form"
      novalidate=""
    >
      <div
        class="px-6 space-y-6"
        data-slot="card-content"
      >
        <div
          class="space-y-2"
        >
          <label
            class="flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50"
            data-slot="label"
            for="email"
          >
            Email
          </label>
          <input
            aria-describedby="email-error"
            aria-invalid="true"
            class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive"
            data-slot="input"
            data-testid="email-input"
            id="email"
            placeholder="name@example.com"
            required=""
            type="email"
            value=""
          />
          <p
            aria-live="polite"
            class="text-sm text-destructive mt-1"
            id="email-error"
            role="alert"
          >
            Email is required
          </p>
        </div>
        <div
          class="space-y-2"
        >
          <label
            class="flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50"
            data-slot="label"
            for="password"
          >
            Password
          </label>
          <input
            aria-describedby="password-error"
            aria-invalid="true"
            class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive"
            data-slot="input"
            data-testid="password-input"
            id="password"
            required=""
            type="password"
            value=""
          />
          <p
            aria-live="polite"
            class="text-sm text-destructive mt-1"
            id="password-error"
            role="alert"
          >
            Password is required
          </p>
        </div>
      </div>
      <div
        class="items-center px-6 [.border-t]:pt-6 flex flex-col space-y-4 pt-6"
        data-slot="card-footer"
      >
        <button
          class="inline-flex items-center justify-center cursor-pointer gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3 w-full"
          data-slot="button"
          data-testid="submit-button"
          type="submit"
        >
          Sign in
        </button>
        <div
          class="flex flex-col space-y-2 text-center text-sm text-muted-foreground"
        >
          <p>
            Don't have an account? 
            <a
              class="inline-flex items-center justify-center cursor-pointer gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive underline-offset-4 h-9 px-4 py-2 has-[>svg]:px-3 text-primary hover:underline"
              data-testid="register-link"
              href="/register"
            >
              Create an account
            </a>
          </p>
          <p>
            <a
              class="inline-flex items-center justify-center cursor-pointer gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive underline-offset-4 h-9 px-4 py-2 has-[>svg]:px-3 text-primary hover:underline"
              data-testid="reset-password-link"
              href="/reset-password"
            >
              Forgot your password?
            </a>
          </p>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
